pyquil/reference_simulator.py

```diff
@@ -165,6 +165,25 @@ def __init__(self, n_qubits: int, rs: RandomState = None):
         self.rs = rs
         self.density = np.zeros((2 ** n_qubits, 2 ** n_qubits), dtype=np.complex128)
         self.density[0, 0] = complex(1.0, 0)
+        self.density_is_set = False
+
+    def set_density(self, state_matrix):
+        """
+        The default state of ReferenceDensitySimulator is |000...00>. This method is the correct
+        way (TM) to set the state matrix to another state.
+
+        :param state_matrix: numpy.ndarray
+        :return: zippo
+        """
+        rows, cols = state_matrix.shape
+        if rows != cols:
+            raise ValueError("The state matrix is not square.")
+        if self.n_qubits != int(np.log2(rows)):
+            raise ValueError("The state matrix is not defined on the same numbers of qubits as "
+                             "the QVM.")
+        self.density = state_matrix
+        self.initial_density = state_matrix
+        self.density_is_set = True
```

Comment:
A couple thoughts here:

1. You could get rid of `self.density_is_set`. To test whether an initial state is provided, simply do `if self.initial_density`.
2. This essentially reimplements a subset of the code in `reset()`. I would argue that maybe this function isn't even needed -- extend `reset(self)` to be `reset(self, initial_state=None)`.

Reply:
`if self.initial_density` is much better, done.

Okay I thought about the other point for a bit. I agree that it is possible to do it all in `reset()` as you suggest. I came up with two good reasons for doing it using `set_density`. 

1. `set_density` makes the action of the user explicit. Related to that, because the method is appropriately named so it will make the code easy to read. 
2. extending `reset` means that function would perform two jobs and only one of them is resetting.

See what you think about the changes I made. I don't feel strongly about this, so I'd be happy to change the code to do it all in `reset` if that is what you (and others if they have opinions) think is best. The change is pretty simple.

Reply:
I originally wrote out a comment here, but it started meandering and feeling condescending. My main gripe is that we have two functions that do similar things, and that usually suggests that we can either (1) make one of those functions more general and remove the other, or (2) remove any duplication by having one function call the other (e.g. `reset(matrix=None)` takes some optional matrix and calls `set_density(matrix)` where matrix is the original matrix (if provided) or the zero state. `set_density(matrix)` would then do some sanity checking, and also update the internal state.

All that to say: who cares. I won't make it a requirement for this PR, unless others agree.

Reply:
I'm inclined towards giving `reset` an optional input as well. It seems to be fairly standard functionality in other APIs I've seen. If the issue is code readability, we could make `_set_density` a private function, and have it optionally called inside `reset`.

Reply:
The issue is not about readability of the internal code it is about the users code being readable.

E.g. 
```python
rho = np.array([[ 0.0, 0.0], [0.0,  1.0]]) #some initial state
qc.qam.wf_simulator.set_density(rho)
qc.run(prog)
```
vs

```python
rho = np.array([[ 0.0, 0.0], [0.0,  1.0]]) #some initial state
qc.qam.wf_simulator.reset(rho)
qc.run(prog)
```
But in light of that argument I would say that I should change the method name to `set_initial_state`.

```python
rho = np.array([[ 0.0, 0.0], [0.0,  1.0]]) #some initial state
qc.qam.wf_simulator.set_initial_state(rho)
qc.run(prog)
```

Anyway both of you think it should be in the reset method. So maybe I should just do that :)

Reply:
In the light of the afternoon, I feel much better about taking @joshcombes' side on this if the function is instead called `set_initial_state`. That more clearly communicates to me what the `reset` function will do.

So: I'm back at square one.

Reply:
IMO it makes the most sense to simply
```qc.qam.wf_simulator.density = state_matrix```
iff you want to change the current state, and separately
```qc.qam.wf_simulator.initial_density = state_matrix``` 
iff you want to specify the state after the next reset. 

I think we should keep these separate. If it's best practice to provide functions to change these variables then we should provide
```set_density(state_matrix)```
and
```set_initial_state(state_matrix)```

Reply:
Because of the discussion above I changed the code a fair bit. Hopefully it is closer to how everyone thinks this feature should be implemented.